crates/bevy_input_focus/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+#![forbid(unsafe_code)]
+#![doc(
+    html_logo_url = "https://bevyengine.org/assets/icon.png",
+    html_favicon_url = "https://bevyengine.org/assets/icon.png"
+)]
+
+//! Keyboard focus system for Bevy.
+//!
+//! This crate provides a system for managing keyboard focus in Bevy applications, including:
+//! * A resource for tracking which entity has keyboard focus.
+//! * Methods for getting and setting keyboard focus.
+//! * Event definitions for triggering bubble-able keyboard input events to the focused entity.
+//! * A system for dispatching keyboard input events to the focused entity.
+//!
+//! This crate does *not* provide any integration with UI widgets, or provide functions for
+//! tab navigation or gamepad-based focus navigation, as those are typically application-specific.
+
+use bevy_app::{App, Plugin, Update};
+use bevy_ecs::{
+    component::Component,
+    entity::Entity,
+    event::{Event, EventReader},
+    query::With,
+    system::{Commands, Query, Res, Resource},
+    world::{Command, DeferredWorld, World},
+};
+use bevy_hierarchy::Parent;
+use bevy_input::keyboard::KeyboardInput;
+use bevy_window::PrimaryWindow;
+
+/// Resource representing which entity has keyboard focus, if any. Keyboard events will be
+/// dispatched to the current focus entity, or to the primary window if no entity has focus.
+#[derive(Clone, Debug, Resource)]
+pub struct KeyboardFocus(pub Option<Entity>);
+
+/// Resource representing whether the keyboard focus indicator should be visible. It's up to the
+/// current focus navigation system to set this resource. For a desktop/web style of user interface
+/// this would be set to true when the user presses the tab key, and set to false when the user
+/// clicks on a different element.
+#[derive(Clone, Debug, Resource)]
+pub struct KeyboardFocusVisible(pub bool);
+
+/// Helper functions for [`World`] and [`DeferredWorld`] to set and clear keyboard focus.
+pub trait SetKeyboardFocus {
+    /// Set keyboard focus to the given entity.
+    fn set_keyboard_focus(&mut self, entity: Entity);
+    /// Clear keyboard focus.
+    fn clear_keyboard_focus(&mut self);
+}
```

Comment:
Should these names be input device agnostic? The keyboards is one of the ways to change focus.

Reply:
Unfortunately, I need *some* sort of adjective here: Bevy uses the word "focus" to mean a couple of different things, for example there is a "focus" concept in picking which has nothing to do with keyboard devices. So we need to be able to distinguish between a navigational focus, and other uses of the word focus.

Slight digression: Bevy's use of the word "focus" is obsolescent - that is, UI frameworks in the early era of personal computers did in fact have a broader definition of the term, but in modern UIs the word "focus" generally means keyboard focus or something like it.

Reply:
Sorry, I might not be understanding this correctly.

I was questioning the `Keyboard` part. Should we drop the `Keyboard` part in these names and leave it as `InputFocus`/`DeviceInputFocus`? It's not only keyboard events we want to forward to the focused widgets, it's any kind of event from an input device, no?

[From the GH discussion](https://github.com/bevyengine/bevy/discussions/15374):

> Simple widgets such as buttons and checkboxes can use the same strategy - the only difference is that they respond to fewer keys. A button will respond to SPACE or ENTER, which is equivalent to a button press. All other events are ignored.

These widgets would also handle a button press to toggle the checkbox or trigger the button. Or, for example, an slider could handle directional input to increase/decrease a value.

Also, if we leave it as `InputFocus` from a "focus navigation" pov the name makes more sense?

Reply:
I agree with you that the naming is an issue, and that using "keyboard" is too restrictive by implication. The problem is that other words are also problematic. The "picking" crate also uses the word "focus" (to mean something different), and as picking also involves "input" and "devices", then `InputFocus` creates a risk of ambiguity.

Reply:
I would like to take the undecorated focus terms here, and rename the picking concept to `PickingFocus` and so on. I think that's much less confusing than any of the alternatives, and this really shouldn't be specific to keyboards.

Reply:
Alice expressed this as "focus cursor" in discord. Maybe we could use `cursor` instead: `InputCursor`? Or a mix of `cursor` + `focus`?

Reply:
Cursor is pretty tied to mice in my mind and sort of ambiguous with pointer. I'd rather we centralize around a specific meaning for focus, and remove the old use in picking. Everything I picking is more about hovering than focus.